Set MPI-executable for ParFlow to $SLURM_SRUN when Slurm

mvhulten opened the pull request:
This is needed when GCC+OpenMPI is used.  This is tested on JURECA.

With Intel+ParaStationMPI this already went right; there is no effect when compiling with Intel LLVM.  It only has effect if ${SLURM_SRUN} is set, which means it is a Slurm system (as is already assumed).

Resolves: #87
Replaces: #88

mvhulten committed:
Set MPI-executable for ParFlow to $SLURM_SRUN when Slurm

This is needed when GCC+OpenMPI is used.  This is tested on JURECA.

With Intel+ParaStationMPI this already went right; there is no effect
when compiling with Intel LLVM.  It only has effect if ${SLURM_SRUN} is
set, which means it is a Slurm system (as is already assumed).

Resolves: #87

kvrigor commented:
How about this solution:

1. [ ] Add `export MPIEXEC_EXECUTABLE=/path/to/srun` in environment file
2. [ ] Modify `cmake/FindMPIFortran.cmake`:
    - [ ] Use the `MPIEXEC_EXECUTABLE` environment setting
    - [x] If no `MPIEXEC_EXECUTABLE` has been set, use the autodetected [`MPIEXEC_EXECUTABLE`](https://cmake.org/cmake/help/latest/module/FindMPI.html#variables-for-locating-mpi)

mvhulten commented:
I think I need to

```cmake
if(JSC)
  set(MPIEXEC_EXECUTABLE "${SLURM_SRUN}")
endif()
```

in `cmake/BuildParFlow.cmake`, because Marvin does not appear to be able to use `srun` (with some PMI API, even though `srun --mpi=list` lists `pmi2`).  I'm getting verification from my Marvin admin colleagues about this…

How do I properly check if I'm on JSC hardware?

kvrigor commented:
> I think I need to
> 
> ```cmake
> if(JSC)
>   set(MPIEXEC_EXECUTABLE "${SLURM_SRUN}")
> endif()
> ```
> 
> in `cmake/BuildParFlow.cmake`, because Marvin does not appear to be able to use `srun` (with some PMI API, even though `srun --mpi=list` lists `pmi2`). I'm getting verification from my Marvin admin colleagues about this…
> 
> How do I properly check if I'm on JSC hardware?

It's best to keep the CMake scripts as system-agnostic as possible. All machine-specific settings should be configured thru the environment files (that's why they exist). In this case you can set `MPIEXEC_EXECUTABLE` in `default.2025.env` or on the env files themselves, and then let `cmake/FindMPIFortran.cmake` detect the user-defined `MPIEXEC_EXECUTABLE`. This way we nicely limit the scope of TSMP2 CMake scripts to compile-related options.

mvhulten committed:
Use srun on JSC machines

mvhulten commented:
I've done this in 30bfbcbcd241223f084719a51aed88a040dccdbb.

But I cannot reproduce the original problem anymore.  I went back to `master` and did a clean compile and several of the parallel Tcl scripts from the test directory all pass.  I'm on the same old commits for TSMP2 and parflow.  I'll look at it later again.

mvhulten committed:
Use correct MPIEXEC_EXECUTABLE

Set this in machine environment file and do not set everywhere to $SLURM_SRUN.
On JSC machines srun is used, but on Marvin mpirun must be used.  The latter
requirement may relate to the missing or incomplete support of PMIx on Marvin.
